src/frontier_detection/src/map.cpp

```diff
@@ -0,0 +1,278 @@
+        mpRawPointCloud.fields[i].name = channel_id[i];
+        mpRawPointCloud.fields[i].offset = i * sizeof(float);
+        mpRawPointCloud.fields[i].count = 1;
+        mpRawPointCloud.fields[i].datatype = sensor_msgs::msg::PointField::FLOAT32;
+    }
+
+    mpRawPointCloud.data.resize(mpRawPointCloud.row_step * mpRawPointCloud.height);
+	mpPointCloudSubscriber = this->create_subscription<MapMsg>(
+			"/orbslam3_mono_node/map_points",
+			10,
+			std::bind(&OctoMapNode::AddNewPointCloudCallback, this, std::placeholders::_1));
+	mpMapPublisher = this->create_publisher<octomap_msgs::msg::Octomap>("~/octomap", 10);
+	mpPointCloudPublisher = this->create_publisher<sensor_msgs::msg::PointCloud2>("~/point_cloud", 10);
+	mpMapPublisherTimer = this->create_wall_timer(
+		std::chrono::milliseconds(100), std::bind(&OctoMapNode::PublishMap, this)
+	);
+	mpThrMsgQueueProcess = new std::thread(&OctoMapNode::UpdateMap, this);	
+}
+
+void OctoMapNode::UpdateMap(){
+	MapMsg::SharedPtr msg;
+	while(true){
+		{
+			std::unique_lock<std::mutex> lock(mpMtxMsgQueue);
+			if(mpMsgQueue.empty()){
+				continue;
+			}
+			else{
+				msg = mpMsgQueue.front();	
+				mpMsgQueue.pop_front();
+			}
+	
+		}
+		octomap::Pointcloud cloud;
+		octomap::point3d point;
+		octomap::pointCloud2ToOctomap(msg->cloud, cloud);
+		// ConvertPointCloud2ToOctomap(msg->cloud, cloud);
+		CreatePoint3dFromTransform(msg->pose, point);
+		mpGlobalMap->UpdateMap(cloud, point);
+	}
+}
+
+void OctoMapNode::UpdateRawPointCloud(sensor_msgs::msg::PointCloud2& cloud) {
+
+    // Verify that both point clouds have the same fields
+    if (cloud.fields != mpRawPointCloud.fields) {
+        RCLCPP_ERROR(this->get_logger(), "PointCloud2 messages have different fields and cannot be combined.");
+		return;
+    }
+
+    // Copy header from the first cloud (modify as needed if frame alignment is required)
+    mpRawPointCloud.header = cloud.header;
+
+    // Set the output point cloud properties
+    mpRawPointCloud.height = 1; // Unordered point cloud
+    mpRawPointCloud.is_bigendian = cloud.is_bigendian;
+    mpRawPointCloud.is_dense = cloud.is_dense;
+    mpRawPointCloud.fields = cloud.fields;
+    mpRawPointCloud.point_step = cloud.point_step;
+    mpRawPointCloud.row_step = 0; // Will be updated later
+
+    // Combine point data
+    mpRawPointCloud.data.reserve(cloud.data.size() + mpRawPointCloud.data.size());
+    mpRawPointCloud.data.insert(mpRawPointCloud.data.end(), cloud.data.begin(), cloud.data.end());
+
+    // Update the total number of points
+    mpRawPointCloud.width = (cloud.width * cloud.height) + (mpRawPointCloud.width * mpRawPointCloud.height);
+    mpRawPointCloud.row_step = mpRawPointCloud.width * mpRawPointCloud.point_step;
+}
+
+void OctoMapNode::AddNewPointCloudCallback(const MapMsg::SharedPtr msg){
+	RCLCPP_DEBUG(this->get_logger(), "Adding point cloud to queue for processing");
+	std::unique_lock<std::mutex> lock(mpMtxMsgQueue);
+	mpMsgQueue.push_back(msg);
```

Comment:
I feel we should be processing the latest point cloud instead of adding points to a queue. Do you have any timing comparison on how long it takes to process and publish 1 point cloud to octomap vs how fast this callback is made? The subscriber itself has a queue which we can tune no right?

Reply:
So the point clouds being published from ORBSLAM3 are not from the Tracking thread, so the point clouds are not realtime. The octomap created as a result, may be a few milli seconds behind realtime always, but is more accurate this way. Due to this design choice, the rate of point cloud publication is highly variable going between 10 to 200Hz. I have not gotten a chance to check how much time is spent updating the octomap per frame. That is something that I have been planning to do. As for the ROS2 queue size, that is currently set at 10. Once I get an average of how time it takes to update the octomap per frame, I will change this value.

Reply:
I was previously using updateInnerOccupancy() which is pretty slow for larger maps. I am now iterating on all map points and setting all nodes manually, which github issues on octomap repo say is the faster method. Even octomap_server does the same. So I am hoping the update times are low only. But I will only be able to say more once I am able to collect more data.

Reply:
ok